### ISSUE TITLE
[CodeView] CommandLineArgs is used after being freed

### ISSUE THREAD

**Issue by aheejin** [llvm/llvm-project#52704]
https://github.com/llvm/llvm-project/commit/89ea0b05207d45c145fb525df554b3b986ae379b added `CommaneLineArgs` to `CodeGenOptions`, which is later [transferred](https://github.com/llvm/llvm-project/blob/ec63930c06da250bc3fbeefdd7e46d18b5d2dc4b/clang/lib/CodeGen/BackendUtil.cpp#L641) to `MCTargetOptions` for later use.

But after https://github.com/llvm/llvm-project/commit/d748908fa02b11c7840a7f03c7a52223126bdba9, this array is freed before use. The origin of that `CommandLineArgs` is `GeneratedArgs1` in the screenshot below. In LHS, `GeneratedArgs1` was an `InputArgList` that contained an array of strings in itself, but in RHS, `GeneratedArgs1` became a stack variable within `RoundTrip` function, which was freed when the function returns:
![image](https://user-images.githubusercontent.com/8726997/146096891-233ad814-0bd0-4928-af24-bc2464c98e1a.png)
https://github.com/llvm/llvm-project/blob/3ce1e9428b96339c516da926657f830bfb4d4c9a/clang/lib/Frontend/CompilerInvocation.cpp#L692

So by the time `CommandLineArgs` is [transferred](https://github.com/llvm/llvm-project/blob/ec63930c06da250bc3fbeefdd7e46d18b5d2dc4b/clang/lib/CodeGen/BackendUtil.cpp#L641) to `MCTargetOptions`, its memory has been already freed and corrupted.

To see how this fails, create these files:

print.diff
```diff
diff --git a/clang/lib/CodeGen/BackendUtil.cpp b/clang/lib/CodeGen/BackendUtil.cpp
index 510f3911939c..638312c9fc92 100644
--- a/clang/lib/CodeGen/BackendUtil.cpp
+++ b/clang/lib/CodeGen/BackendUtil.cpp
@@ -639,6 +639,8 @@ static bool initTargetOptions(DiagnosticsEngine &Diags,
           Entry.IgnoreSysRoot ? Entry.Path : HSOpts.Sysroot + Entry.Path);
   Options.MCOptions.Argv0 = CodeGenOpts.Argv0;
   Options.MCOptions.CommandLineArgs = CodeGenOpts.CommandLineArgs;
+  for (const char *s : Options.MCOptions.CommandLineArgs)
+    llvm::errs() << s << "\n";
   Options.DebugStrictDwarf = CodeGenOpts.DebugStrictDwarf;
 
   return true;
```

test.ll
```
define i32 @main() {
ret:
  ret i32 0
}
```

Patch the clang+llvm codebase with print.diff file:
```
(In llvm-project top directory)
$ patch -p1 < print.diff
```

Build the codebase with ASan enabled, and run the command below:
```
$ clang-12 -cc1 -emit-obj -round-trip-args test.ll
```
Or build the codebase with ASan+assert enabled, and run the command below:
```
$ clang++ test.ll -c
```
(The second command may not fail before https://github.com/llvm/llvm-project/commit/3950e1be8d6ec602fda525ea20340fd3f8e2e919, which mandated the roundtripping for assert builds.)

Then the error message is:
```
=================================================================
==189302==ERROR: AddressSanitizer: heap-use-after-free on address 0x60b000000ca0 at pc 0x7f038590b548 bp 0x7ffe75cdb850 sp 0x7ffe75cdb848
READ of size 8 at 0x60b000000ca0 thread T0
    #0 0x7f038590b547 in initTargetOptions(clang::DiagnosticsEngine&, llvm::TargetOptions&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, clang::HeaderSearchOptions const&) /usr/local/google/home/aheejin/llvm-git/clang/lib/CodeGen/BackendUtil.cpp:642:22
    #1 0x7f038590cf8b in (anonymous namespace)::EmitAssemblyHelper::CreateTargetMachine(bool) /usr/local/google/home/aheejin/llvm-git/clang/lib/CodeGen/BackendUtil.cpp:947:8
    #2 0x7f03859007bf in (anonymous namespace)::EmitAssemblyHelper::EmitAssembly(clang::BackendAction, std::__1::unique_ptr<llvm::raw_pwrite_stream, std::__1::default_delete<llvm::raw_pwrite_stream> >) /usr/local/google/home/aheejin/llvm-git/clang/lib/CodeGen/BackendUtil.cpp:1542:3
    #3 0x7f03859007bf in clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, std::__1::unique_ptr<llvm::raw_pwrite_stream, std::__1::default_delete<llvm::raw_pwrite_stream> >) /usr/local/google/home/aheejin/llvm-git/clang/lib/CodeGen/BackendUtil.cpp:1716:15
    #4 0x7f0386404c32 in clang::CodeGenAction::ExecuteAction() /usr/local/google/home/aheejin/llvm-git/clang/lib/CodeGen/CodeGenAction.cpp:1175:3
    #5 0x7f0380368452 in clang::FrontendAction::Execute() /usr/local/google/home/aheejin/llvm-git/clang/lib/Frontend/FrontendAction.cpp:967:8
    #6 0x7f0380219516 in clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) /usr/local/google/home/aheejin/llvm-git/clang/lib/Frontend/CompilerInstance.cpp:1030:33
    #7 0x7f0386e378b5 in clang::ExecuteCompilerInvocation(clang::CompilerInstance*) /usr/local/google/home/aheejin/llvm-git/clang/lib/FrontendTool/ExecuteCompilerInvocation.cpp:261:25
    #8 0x337900 in cc1_main(llvm::ArrayRef<char const*>, char const*, void*) /usr/local/google/home/aheejin/llvm-git/clang/tools/driver/cc1_main.cpp:246:15
    #9 0x32c3be in ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&) /usr/local/google/home/aheejin/llvm-git/clang/tools/driver/driver.cpp:317:12
    #10 0x32799d in main /usr/local/google/home/aheejin/llvm-git/clang/tools/driver/driver.cpp:388:12
    #11 0x7f0376d58e49 in __libc_start_main csu/../csu/libc-start.c:314:16
    #12 0x275b89 in _start (/usr/local/google/home/aheejin/llvm-git/install.asan/bin/clang-12+0x275b89)

0x60b000000ca0 is located 0 bytes inside of 104-byte region [0x60b000000ca0,0x60b000000d08)
freed by thread T0 here:
    #0 0x2f2472 in free /usr/local/google/home/aheejin/apps/llvm-project-13.0.0/compiler-rt/lib/asan/asan_malloc_linux.cpp:111:3
    #1 0x7f03803380c4 in llvm::SmallVectorImpl<char const*>::~SmallVectorImpl() /usr/local/google/home/aheejin/llvm-git/llvm/include/llvm/ADT/SmallVector.h:582:7
    #2 0x7f03803380c4 in llvm::SmallVector<char const*, 6u>::~SmallVector() /usr/local/google/home/aheejin/llvm-git/llvm/include/llvm/ADT/SmallVector.h:1186:3
    #3 0x7f03802e096e in RoundTrip(llvm::function_ref<bool (clang::CompilerInvocation&, llvm::ArrayRef<char const*>, clang::DiagnosticsEngine&, char const*)>, llvm::function_ref<void (clang::CompilerInvocation&, llvm::SmallVectorImpl<char const*>&, llvm::function_ref<char const* (llvm::Twine const&)>)>, clang::CompilerInvocation&, clang::CompilerInvocation&, llvm::ArrayRef<char const*>, clang::DiagnosticsEngine&, char const*) /usr/local/google/home/aheejin/llvm-git/clang/lib/Frontend/CompilerInvocation.cpp:742:1
    #4 0x7f03802e096e in clang::CompilerInvocation::CreateFromArgs(clang::CompilerInvocation&, llvm::ArrayRef<char const*>, clang::DiagnosticsEngine&, char const*) /usr/local/google/home/aheejin/llvm-git/clang/lib/Frontend/CompilerInvocation.cpp:4536:10
    #5 0x337225 in cc1_main(llvm::ArrayRef<char const*>, char const*, void*) /usr/local/google/home/aheejin/llvm-git/clang/tools/driver/cc1_main.cpp:212:18
    #6 0x32c3be in ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&) /usr/local/google/home/aheejin/llvm-git/clang/tools/driver/driver.cpp:317:12
    #7 0x32799d in main /usr/local/google/home/aheejin/llvm-git/clang/tools/driver/driver.cpp:388:12
    #8 0x7f0376d58e49 in __libc_start_main csu/../csu/libc-start.c:314:16

previously allocated by thread T0 here:
    #0 0x2f26dd in malloc /usr/local/google/home/aheejin/apps/llvm-project-13.0.0/compiler-rt/lib/asan/asan_malloc_linux.cpp:129:3
    #1 0x7f0377701a69 in llvm::safe_malloc(unsigned long) /usr/local/google/home/aheejin/llvm-git/llvm/include/llvm/Support/MemAlloc.h:26:18
    #2 0x7f0377701a69 in llvm::SmallVectorBase<unsigned int>::grow_pod(void*, unsigned long, unsigned long) /usr/local/google/home/aheejin/llvm-git/llvm/lib/Support/SmallVector.cpp:125:15
    #3 0x7f038033a640 in llvm::SmallVectorTemplateCommon<char const*, void>::grow_pod(unsigned long, unsigned long) /usr/local/google/home/aheejin/llvm-git/llvm/include/llvm/ADT/SmallVector.h:124:11
    #4 0x7f038033a640 in llvm::SmallVectorTemplateBase<char const*, true>::grow(unsigned long) /usr/local/google/home/aheejin/llvm-git/llvm/include/llvm/ADT/SmallVector.h:510:41
    #5 0x7f038033a5dc in char const* const* llvm::SmallVectorTemplateCommon<char const*, void>::reserveForParamAndGetAddressImpl<llvm::SmallVectorTemplateBase<char const*, true> >(llvm::SmallVectorTemplateBase<char const*, true>*, char const* const&, unsigned long) /usr/local/google/home/aheejin/llvm-git/llvm/include/llvm/ADT/SmallVector.h:229:11
    #6 0x7f038030d0e3 in llvm::SmallVectorTemplateBase<char const*, true>::reserveForParamAndGetAddress(char const*&, unsigned long) /usr/local/google/home/aheejin/llvm-git/llvm/include/llvm/ADT/SmallVector.h:522:9
    #7 0x7f038030d0e3 in llvm::SmallVectorTemplateBase<char const*, true>::push_back(char const*) /usr/local/google/home/aheejin/llvm-git/llvm/include/llvm/ADT/SmallVector.h:547:23
    #8 0x7f03802580fe in denormalizeSimpleFlag(llvm::SmallVectorImpl<char const*>&, char const*, llvm::function_ref<char const* (llvm::Twine const&)>, llvm::opt::Option::OptionClass, unsigned int, ...) /usr/local/google/home/aheejin/llvm-git/clang/lib/Frontend/CompilerInvocation.cpp:181:8
    #9 0x7f03802580fe in GenerateArg(llvm::SmallVectorImpl<char const*>&, llvm::opt::OptSpecifier, llvm::function_ref<char const* (llvm::Twine const&)>) /usr/local/google/home/aheejin/llvm-git/clang/lib/Frontend/CompilerInvocation.cpp:591:3
    #10 0x7f0380248952 in clang::CompilerInvocation::GenerateCodeGenArgs(clang::CodeGenOptions const&, llvm::SmallVectorImpl<char const*>&, llvm::function_ref<char const* (llvm::Twine const&)>, llvm::Triple const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, clang::LangOptions const*) /usr/local/google/home/aheejin/llvm-git/clang/lib/Frontend/CompilerInvocation.cpp:1319:5
    #11 0x7f03802f35fd in clang::CompilerInvocation::generateCC1CommandLine(llvm::SmallVectorImpl<char const*>&, llvm::function_ref<char const* (llvm::Twine const&)>) const /usr/local/google/home/aheejin/llvm-git/clang/lib/Frontend/CompilerInvocation.cpp:4654:3
    #12 0x7f0380308509 in clang::CompilerInvocation::CreateFromArgs(clang::CompilerInvocation&, llvm::ArrayRef<char const*>, clang::DiagnosticsEngine&, char const*)::$_775::operator()(clang::CompilerInvocation&, llvm::SmallVectorImpl<char const*>&, llvm::function_ref<char const* (llvm::Twine const&)>) const /usr/local/google/home/aheejin/llvm-git/clang/lib/Frontend/CompilerInvocation.cpp:4542:43
    #13 0x7f0380308509 in void llvm::function_ref<void (clang::CompilerInvocation&, llvm::SmallVectorImpl<char const*>&, llvm::function_ref<char const* (llvm::Twine const&)>)>::callback_fn<clang::CompilerInvocation::CreateFromArgs(clang::CompilerInvocation&, llvm::ArrayRef<char const*>, clang::DiagnosticsEngine&, char const*)::$_775>(long, clang::CompilerInvocation&, llvm::SmallVectorImpl<char const*>&, llvm::function_ref<char const* (llvm::Twine const&)>) /usr/local/google/home/aheejin/llvm-git/llvm/include/llvm/ADT/STLExtras.h:177:12
    #14 0x7f0380337fbd in llvm::function_ref<void (clang::CompilerInvocation&, llvm::SmallVectorImpl<char const*>&, llvm::function_ref<char const* (llvm::Twine const&)>)>::operator()(clang::CompilerInvocation&, llvm::SmallVectorImpl<char const*>&, llvm::function_ref<char const* (llvm::Twine const&)>) const /usr/local/google/home/aheejin/llvm-git/llvm/include/llvm/ADT/STLExtras.h:200:12
    #15 0x7f03802dfb83 in RoundTrip(llvm::function_ref<bool (clang::CompilerInvocation&, llvm::ArrayRef<char const*>, clang::DiagnosticsEngine&, char const*)>, llvm::function_ref<void (clang::CompilerInvocation&, llvm::SmallVectorImpl<char const*>&, llvm::function_ref<char const* (llvm::Twine const&)>)>, clang::CompilerInvocation&, clang::CompilerInvocation&, llvm::ArrayRef<char const*>, clang::DiagnosticsEngine&, char const*) /usr/local/google/home/aheejin/llvm-git/clang/lib/Frontend/CompilerInvocation.cpp:693:3
    #16 0x7f03802dfb83 in clang::CompilerInvocation::CreateFromArgs(clang::CompilerInvocation&, llvm::ArrayRef<char const*>, clang::DiagnosticsEngine&, char const*) /usr/local/google/home/aheejin/llvm-git/clang/lib/Frontend/CompilerInvocation.cpp:4536:10
    #17 0x337225 in cc1_main(llvm::ArrayRef<char const*>, char const*, void*) /usr/local/google/home/aheejin/llvm-git/clang/tools/driver/cc1_main.cpp:212:18
    #18 0x32c3be in ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&) /usr/local/google/home/aheejin/llvm-git/clang/tools/driver/driver.cpp:317:12
    #19 0x32799d in main /usr/local/google/home/aheejin/llvm-git/clang/tools/driver/driver.cpp:388:12
    #20 0x7f0376d58e49 in __libc_start_main csu/../csu/libc-start.c:314:16

SUMMARY: AddressSanitizer: heap-use-after-free /usr/local/google/home/aheejin/llvm-git/clang/lib/CodeGen/BackendUtil.cpp:642:22 in initTargetOptions(clang::DiagnosticsEngine&, llvm::TargetOptions&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, clang::HeaderSearchOptions const&)
Shadow bytes around the buggy address:
  0x0c167fff8140: 00 00 00 00 00 00 00 00 00 fa fa fa fa fa fa fa
  0x0c167fff8150: fa fa 00 00 00 00 00 00 00 00 00 00 00 00 00 fa
  0x0c167fff8160: fa fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
  0x0c167fff8170: 00 00 00 00 00 fa fa fa fa fa fa fa fa fa 00 00
  0x0c167fff8180: 00 00 00 00 00 00 00 00 00 00 00 00 fa fa fa fa
=>0x0c167fff8190: fa fa fa fa[fd]fd fd fd fd fd fd fd fd fd fd fd
  0x0c167fff81a0: fd fa fa fa fa fa fa fa fa fa fd fd fd fd fd fd
  0x0c167fff81b0: fd fd fd fd fd fd fd fa fa fa fa fa fa fa fa fa
  0x0c167fff81c0: 00 00 00 00 00 00 00 00 00 00 00 00 01 fa fa fa
  0x0c167fff81d0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c167fff81e0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==189302==ABORTING
```

I'm not very sure if this feature is still in use. I discovered this by chance while working on something else.
cc @aganea @jansvoboda11

**Comment by aganea** [comment]
Thanks for the report! I can take a look.

**Comment by aganea** [comment]
Fixed by d26520f6f78785b0c4c296a8a992f2adb656c6ec
